### PR TITLE
feat(shell-api): add stream processor modify MONGOSH-1864

### PR DESCRIPTION
**Pull request by mongodb-matthew-normyle**
This PR adds a stream processor `modify` method. It can be used to modify the definition of a stream processor.

### CODE REVIEW

**Comment by mongodb-matthew-normyle** [comment]
I also manually tested this change against the ASP service with the below:

```
AtlasStreamProcessing> sp.createStreamProcessor("testedit2", [{$source: {connectionName: "DemoKafka", topic: "t1"}}, {$project: {device_id: 1, foo: "bar"}}, {$project: {_stream_meta: 0}}, {$emit: {connectionName: "DemoKafka", topic: "testedit2"}}])
Atlas Stream Processor: testedit2
AtlasStreamProcessing> sp.testedit2.start()
{ ok: 1 }
AtlasStreamProcessing> sp.testedit2.stop()
{ ok: 1 }
AtlasStreamProcessing> sp.testedit2.modify([{$source: {connectionName: "DemoKafka", topic: "t1"}}, {$project: {device_id: 1, foo: "BAZ"}}, {$project: {_stream_meta: 0}}, {$emit: {connectionName: "DemoKafka", topic: "testedit2"}}])
{ ok: 1 }
AtlasStreamProcessing> sp.testedit2.start()
{ ok: 1 }

```

File: packages/shell-api/src/stream-processor.ts
Comment (mongodb-matthew-normyle) [inline packages/shell-api/src/stream-processor.ts:73]: The method parameters were defined in the [Scope](https://docs.google.com/document/d/1wQ_e2TrJKegyei0nDFajVejE6nqahmftieyZ3UEMnXE/edit?tab=t.0) and [PD](https://docs.google.com/document/d/1Qvw4fI6DrSxqJbhODnHaY1p3aJIz0632KsJ-qK401d8/edit?tab=t.0#heading=h.wdpdgdpcthgg). I think [watch](https://www.mongodb.com/docs/manual/reference/method/db.collection.watch/) has a similar structure that support arguments of different types. But if this is against the best mongosh practices we can modify the scope.

**Comment by mongodb-matthew-normyle** [comment]
I'm waiting on an evergreen build to see if there are any other linting errors to fix.

File: packages/shell-api/src/stream-processor.ts
Comment (addaleax) [reply packages/shell-api/src/stream-processor.ts:73]: This is fine, but:
- Please add TS overloads that describe the different signatures, i.e. `modify(options: Document): ...` and `modify(pipeline: Document[], options: Document): ...`
- Please add return types to the definitions

File: packages/shell-api/src/stream-processor.ts
Comment (addaleax) [inline packages/shell-api/src/stream-processor.ts:76]:
```suggestion
      options.pipeline = pipelineOrOptions;
```

or, even better:

```suggestion
      options = { ...options, pipeline: pipelineOrOptions };
```

since best practice is not to modify arguments passed to functions

File: packages/shell-api/src/stream-processor.ts
Comment (addaleax) [inline packages/shell-api/src/stream-processor.ts:78]:
```suggestion
      if (options) {
```

File: packages/shell-api/src/stream-processor.ts
Comment (mongodb-matthew-normyle) [reply packages/shell-api/src/stream-processor.ts:78]: Done, thanks!

File: packages/shell-api/src/stream-processor.ts
Comment (mongodb-matthew-normyle) [reply packages/shell-api/src/stream-processor.ts:83]: I think `if (options) {}` evaluates to true when options is `{}` (the default argument). So, if I make this change,  I'll need to change `options: Document = {}` to `options: Document = null`. Do you have a preference? I think keeping `options: Document = {}` and `if (Object.keys(options).length !== 0) ` is slightly better.

Apologies if I'm missing something-- this is the first typescript PR I've written :)

File: packages/shell-api/src/stream-processor.ts
Comment (mongodb-matthew-normyle) [reply packages/shell-api/src/stream-processor.ts]: Thanks, done.

File: packages/shell-api/src/stream-processor.ts
Comment (addaleax) [inline packages/shell-api/src/stream-processor.ts:68]:
```suggestion
  async modify(pipelineOrOptions: Document[] | Document): Promise<Document>;
  async modify(pipeline: Document[], options: Document): Promise<Document>;
```

File: packages/shell-api/src/stream-processor.ts
Comment (addaleax) [reply packages/shell-api/src/stream-processor.ts:83]: Yeah, the `options: Document = {}` above would become `options?: Document` in this case 



> I think keeping `options: Document = {}` and `if (Object.keys(options).length !== 0) ` is slightly better.

I'd say checking the # of keys of an argument to check whether it was passed or not is definitely a bit unidiomatic. For example, it will just throw an unexpected exception if somebody passes `null` as an option objects.

Comment (mongodb-matthew-normyle) [reply packages/shell-api/src/stream-processor.ts]: Ah yes that is much better, thanks. I didn't realize that options?: Document would work with the spread operator in `{...options, pipeline: pipelineOrOptions }`.

**Review by addaleax — APPROVED** [review]

**Comment by addaleax** [comment]
@mongodb-matthew-normyle The outstanding CI failures are known flaky/failing tests, so this would be good to merge if you're read 👍

**Comment by mongodb-matthew-normyle** [comment]
> @mongodb-matthew-normyle The outstanding CI failures are known flaky/failing tests, so this would be good to merge if you're read 👍

Hi @addaleax, I'm ready to merge :) It appears I don't have write access to merge my PR. Are you able to hit merge for me? Thanks in advance.

**Pull request merged**
